File: Java/11-ContainerWithMostWater.java
public class Solution {
    public int maxArea(int[] height) {
        if (height == null || height.length == 0){
            return 0;
        }
        int left = 0, right = height.length - 1, maxA = 0;
        while (left < right){
            maxA = Math.max(Math.min(height[left], height[right]) * (right - left), maxA);
            if (height[left] >= height[right] && left < right){
                right--;
            }else if (height[left] < height[right] && left < right){
                left++;
            }
        }
        return maxA;
    }
}

File: cpp/11-ContainerWithMostWater.cpp
//[1,2,3,2,4,2,1]
class Solution {
public:
    int maxArea(vector<int>& height) {
        int maxWater = 0, left = 0, right = height.size() - 1;
        while (left < right) {
            int currWater = (right - left) * min(height[left], height[right]);
            maxWater = max(maxWater, currWater);
            if (height[left] <= height[right]) {
                left++;
            } else {
                right--;
            }
        }
        return maxWater;
    }
};
